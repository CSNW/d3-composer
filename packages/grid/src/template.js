import { toMargin } from '@d3-composer/utils';
import { parseTemplate, parseRows, parseColumns, parseAreas } from './parse';
import solve from './solve';

const cache = new Map();

export default function template(spec, options) {
  let { width, height, row_gap, column_gap, padding } = options;
  padding = toMargin(padding);
  width -= padding[1] + padding[3];
  height -= padding[0] + padding[2];

  // spec: string = template
  // spec: object = { rows, columns, [areas] } templates
  let rows, columns, areas;
  if (typeof spec === 'object') {
    rows = parseRows(spec.rows);
    columns = parseColumns(spec.columns);
    areas = spec.areas ? parseAreas(spec.areas) : {};
  } else {
    const parsed = cache.has(spec) ? cache.get(spec) : parseTemplate(spec);
    cache.set(spec, parsed);
    ({ rows, columns, areas } = parsed);
  }

  const x = solve(columns, width, column_gap, padding[3]);
  const y = solve(rows, height, row_gap, padding[0]);

  return layout(rows, columns, areas, x, y);
}

export function layout(rows, columns, areas, x, y) {
  const result = { rows: y, columns: x, items: [], areas: {} };

  for (let i = 0; i < rows.length; i++) {
    for (let j = 0; j < columns.length; j++) {
      const top = y[i][0];
      const right = x[j][1];
      const bottom = y[i][1];
      const left = x[j][0];

      result.items.push({
        x: left,
        y: top,
        width: right - left,
        height: bottom - top,
        top,
        right,
        bottom,
        left
      });
    }
  }

  for (const name of Object.keys(areas)) {
    const area = areas[name];

    const top = y[area.top][0];
    const right = x[area.right][1];
    const bottom = y[area.bottom][1];
    const left = x[area.left][0];

    result.areas[name] = {
      x: left,
      y: top,
      width: right - left,
      height: bottom - top,
      top,
      right,
      bottom,
      left
    };
  }

  return result;
}
