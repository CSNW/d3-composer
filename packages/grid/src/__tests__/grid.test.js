import template from '../grid';

test('should create grid from spec and size', () => {
  const spec = `
    "title title" 50
    "y_axis chart" auto
    ". x_axis" 50
    / 50 auto
  `;

  expect(template(spec, { width: 600, height: 400 })).toMatchSnapshot();
});
