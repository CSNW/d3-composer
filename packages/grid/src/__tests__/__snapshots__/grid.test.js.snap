// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create grid from spec and size 1`] = `
Object {
  "chart": Object {
    "bottom": 350,
    "height": 300,
    "left": 50,
    "right": 600,
    "top": 50,
    "width": 550,
  },
  "title": Object {
    "bottom": 50,
    "height": 50,
    "left": 0,
    "right": 600,
    "top": 0,
    "width": 600,
  },
  "x_axis": Object {
    "bottom": 400,
    "height": 50,
    "left": 50,
    "right": 600,
    "top": 350,
    "width": 550,
  },
  "y_axis": Object {
    "bottom": 350,
    "height": 300,
    "left": 0,
    "right": 50,
    "top": 50,
    "width": 50,
  },
}
`;
